### PR TITLE
[3.x] Fix game controllers ignoring the last listed button

### LINKED ISSUE #47579: Home button on Xbox-like controllers is not recognized
**Godot version:**
3.2.4 rc3 official

**OS/device including version:**
Windows 10, 64 bit

**Issue description:**
The HOME button on a Xbox like controller (XInput, like the Logitech F310) is not in the [JoystickList](https://docs.godotengine.org/en/stable/classes/class_%40globalscope.html#enum-globalscope-joysticklist).

Also, when pressing the HOME button on a Xbox like controller, it returns the scancode 33554431 (Unknown key).

I think this is a fault of the engine, because platforms like Steam expect a game to have an answer to the HOME button, like pausing the game. It can be done with Steam API integration, but why not recognize the button?

Users also expect this, because even if it's not on a specific store/launch hub, on Windows, XInput controllers will open the Xbox Game Bar with the HOME button, and it is good practice to pause the game when that happens.

**Steps to reproduce:**
Just add a script to any project with the code below, and press the HOME button in a Xbox like controller.

```
func _unhandled_input(event):
	if event is InputEventKey:
		print(event.scancode)
```

**Minimal reproduction project:**
See above.

### PR DESCRIPTION
**Pull request by madmiraal**
Fixes #47579.
Addresses https://github.com/godotengine/godot-demo-projects/pull/597#issuecomment-798826879.

Godot is missing the last listed button on the joypad. Until #45798, this was the `guide` button, a.k.a. the Home button on Xbox controllers and the PS button on DualShock controllers. With #45798 (included in 3.3) this is now the `touchpad` button. However, it also means that the new `enum`s for `misc1`, `paddle1`, `paddle2`, `paddle3`, `paddle4`, and `touchpad` introduced in #47598 are wrong too.

This PR increases the value for `JOY_BUTTON_MAX` to 23, adds the missing `enum` and corrects the values for the `enum`s added in #45798. Technically this breaks compat, but, since the `enum`s are new, no more so than #45798 itself and the [Joypads Demo](https://github.com/godotengine/godot-demo-projects/tree/master/misc/joypads) project was updated to support it already in https://github.com/godotengine/godot-demo-projects/pull/597; so probably worth cherry-picking for 3.3.2.

Note: This was fixed in 4.0 in #38151.

### CODE REVIEW

**Comment by madmiraal** [comment]
For reference, this will be the mapping between SDL outputs and the enum values with this PR:
SDL | Primary enum | Secondary enum
-- | -- | --
a | JOY_BUTTON_0 = 0, | &nbsp;
b | JOY_BUTTON_1 = 1, | &nbsp;
x | JOY_BUTTON_2 = 2, | &nbsp;
y | JOY_BUTTON_3 = 3, | &nbsp;
leftshoulder | JOY_BUTTON_4 = 4, | JOY_L = JOY_BUTTON_4,
rightshoulder | JOY_BUTTON_5 = 5, | JOY_R = JOY_BUTTON_5,
lefttrigger | JOY_BUTTON_6 = 6, | JOY_L2 = JOY_BUTTON_6,
righttrigger | JOY_BUTTON_7 = 7, | JOY_R2 = JOY_BUTTON_7,
leftstick | JOY_BUTTON_8 = 8, | JOY_L3 = JOY_BUTTON_8,
rightstick | JOY_BUTTON_9 = 9, | JOY_R3 = JOY_BUTTON_9,
back | JOY_BUTTON_10 = 10, | JOY_SELECT = JOY_BUTTON_10,
start | JOY_BUTTON_11 = 11, | JOY_START = JOY_BUTTON_11,
dpup | JOY_BUTTON_12 = 12, | JOY_DPAD_UP = JOY_BUTTON_12,
dpdown | JOY_BUTTON_13 = 13, | JOY_DPAD_DOWN = JOY_BUTTON_13,
dpleft | JOY_BUTTON_14 = 14, | JOY_DPAD_LEFT = JOY_BUTTON_14,
dpright | JOY_BUTTON_15 = 15, | JOY_DPAD_RIGHT = JOY_BUTTON_15,
guide | JOY_BUTTON_16 = 16, | JOY_GUIDE = JOY_BUTTON_16,
misc1 | JOY_BUTTON_17 = 17, | JOY_MISC1 = JOY_BUTTON_17,
paddle1 | JOY_BUTTON_18 = 18, | JOY_PADDLE1 = JOY_BUTTON_18,
paddle2 | JOY_BUTTON_19 = 19, | JOY_PADDLE2 = JOY_BUTTON_19,
paddle3 | JOY_BUTTON_20 = 20, | JOY_PADDLE3 = JOY_BUTTON_20,
paddle4 | JOY_BUTTON_21 = 21, | JOY_PADDLE4 = JOY_BUTTON_21,
touchpad | JOY_BUTTON_22 = 22, | JOY_TOUCHPAD = JOY_BUTTON_22,
nullptr | JOY_BUTTON_MAX = 23, | &nbsp;

This is the current mapping:
SDL | Primary enum | Secondary enum
-- | -- | --
a | JOY_BUTTON_0 = 0, | &nbsp;
b | JOY_BUTTON_1 = 1, | &nbsp;
x | JOY_BUTTON_2 = 2, | &nbsp;
y | JOY_BUTTON_3 = 3, | &nbsp;
leftshoulder | JOY_BUTTON_4 = 4, | JOY_L = JOY_BUTTON_4,
rightshoulder | JOY_BUTTON_5 = 5, | JOY_R = JOY_BUTTON_5,
lefttrigger | JOY_BUTTON_6 = 6, | JOY_L2 = JOY_BUTTON_6,
righttrigger | JOY_BUTTON_7 = 7, | JOY_R2 = JOY_BUTTON_7,
leftstick | JOY_BUTTON_8 = 8, | JOY_L3 = JOY_BUTTON_8,
rightstick | JOY_BUTTON_9 = 9, | JOY_R3 = JOY_BUTTON_9,
back | JOY_BUTTON_10 = 10, | JOY_SELECT = JOY_BUTTON_10,
start | JOY_BUTTON_11 = 11, | JOY_START = JOY_BUTTON_11,
dpup | JOY_BUTTON_12 = 12, | JOY_DPAD_UP = JOY_BUTTON_12,
dpdown | JOY_BUTTON_13 = 13, | JOY_DPAD_DOWN = JOY_BUTTON_13,
dpleft | JOY_BUTTON_14 = 14, | JOY_DPAD_LEFT = JOY_BUTTON_14,
dpright | JOY_BUTTON_15 = 15, | JOY_DPAD_RIGHT = JOY_BUTTON_15,
guide | JOY_BUTTON_16 = 16, | JOY_MISC1 = JOY_BUTTON_16,
misc1 | JOY_BUTTON_17 = 17, | JOY_PADDLE1 = JOY_BUTTON_17,
paddle1 | JOY_BUTTON_18 = 18, | JOY_PADDLE2 = JOY_BUTTON_18,
paddle2 | JOY_BUTTON_19 = 19, | JOY_PADDLE3 = JOY_BUTTON_19,
paddle3 | JOY_BUTTON_20 = 20, | JOY_PADDLE4 = JOY_BUTTON_20,
paddle4 | JOY_BUTTON_21 = 21, | JOY_TOUCHPAD = JOY_BUTTON_21,
touchpad | JOY_BUTTON_MAX = 22, | &nbsp;
nullptr | &nbsp; | &nbsp;

This was the mapping prior to #45798:
SDL | Primary enum | Secondary enum
-- | -- | --
a | JOY_BUTTON_0 = 0, | &nbsp;
b | JOY_BUTTON_1 = 1, | &nbsp;
x | JOY_BUTTON_2 = 2, | &nbsp;
y | JOY_BUTTON_3 = 3, | &nbsp;
leftshoulder | JOY_BUTTON_4 = 4, | JOY_L = JOY_BUTTON_4,
rightshoulder | JOY_BUTTON_5 = 5, | JOY_R = JOY_BUTTON_5,
lefttrigger | JOY_BUTTON_6 = 6, | JOY_L2 = JOY_BUTTON_6,
righttrigger | JOY_BUTTON_7 = 7, | JOY_R2 = JOY_BUTTON_7,
leftstick | JOY_BUTTON_8 = 8, | JOY_L3 = JOY_BUTTON_8,
rightstick | JOY_BUTTON_9 = 9, | JOY_R3 = JOY_BUTTON_9,
back | JOY_BUTTON_10 = 10, | JOY_SELECT = JOY_BUTTON_10,
start | JOY_BUTTON_11 = 11, | JOY_START = JOY_BUTTON_11,
dpup | JOY_BUTTON_12 = 12, | JOY_DPAD_UP = JOY_BUTTON_12,
dpdown | JOY_BUTTON_13 = 13, | JOY_DPAD_DOWN = JOY_BUTTON_13,
dpleft | JOY_BUTTON_14 = 14, | JOY_DPAD_LEFT = JOY_BUTTON_14,
dpright | JOY_BUTTON_15 = 15, | JOY_DPAD_RIGHT = JOY_BUTTON_15,
guide | JOY_BUTTON_MAX = 16, | &nbsp;
NULL | &nbsp; | &nbsp;

This is the mapping on master:
SDL | enum
-- | --
a | JOY_BUTTON_A = 0,
b | JOY_BUTTON_B = 1,
x | JOY_BUTTON_X = 2,
y | JOY_BUTTON_Y = 3,
back | JOY_BUTTON_BACK = 4,
guide | JOY_BUTTON_GUIDE = 5,
start | JOY_BUTTON_START = 6,
leftstick | JOY_BUTTON_LEFT_STICK = 7,
rightstick | JOY_BUTTON_RIGHT_STICK = 8,
leftshoulder | JOY_BUTTON_LEFT_SHOULDER = 9,
rightshoulder | JOY_BUTTON_RIGHT_SHOULDER = 10,
dpup | JOY_BUTTON_DPAD_UP = 11,
dpdown | JOY_BUTTON_DPAD_DOWN = 12,
dpleft | JOY_BUTTON_DPAD_LEFT = 13,
dpright | JOY_BUTTON_DPAD_RIGHT = 14,
misc1 | JOY_BUTTON_MISC1 = 15,
paddle1 | JOY_BUTTON_PADDLE1 = 16,
paddle2 | JOY_BUTTON_PADDLE2 = 17,
paddle3 | JOY_BUTTON_PADDLE3 = 18,
paddle4 | JOY_BUTTON_PADDLE4 = 19,
touchpad | JOY_BUTTON_TOUCHPAD = 20,
&nbsp; | JOY_BUTTON_SDL_MAX = 21,
&nbsp; | JOY_BUTTON_MAX = 36,

File: core/global_constants.cpp [hunk 1]
```diff
@@ -442,6 +443,7 @@ void register_global_constants() {
 
 	BIND_GLOBAL_ENUM_CONSTANT(JOY_SELECT);
 	BIND_GLOBAL_ENUM_CONSTANT(JOY_START);
+	BIND_GLOBAL_ENUM_CONSTANT(JOY_GUIDE);
```
Comment (akien-mga) [inline core/global_constants.cpp:446]: This should be after `JOY_DPAD_RIGHT` to respect the order (so `16` comes after `15` and not `11` in the docs).

**Review by akien-mga — APPROVED** [review]
Aside from my comment, seems good :+1:

File: core/global_constants.cpp (hunk 1 above)
Comment (madmiraal) [reply core/global_constants.cpp]: Done.

**Comment by akien-mga** [comment]
Thanks!

**Comment by akien-mga** [comment]
Cherry-picked for 3.3.3.

**Pull request merged**